src/pool.jl

```diff
@@ -1,3 +1,7 @@
+const catpool_seed = UInt === UInt32 ? 0xe3cf1386 : 0x356f2c715023f1a5
+
+hashlevels(levs::AbstractVector) = foldl((h, x) -> hash(x, h), levs, init=catpool_seed)
```

Comment:
Computing the hash isn't actually required to benefit from most of the performance gains of this PR. That's because the hash only allows knowing that two pools are not equal (never that they are equal), but that case is already slow anyway. It implemented `Base.hash` mainly because it's good to have if we implement `==`/`isequal`.

Another point to discuss is what's the best hashing algorithm. The approach I took here is to hash all levels. This is slow on creation, but it's cheap to compute the new hash when adding a level at the end using `hash(newlevel, pool.hash)`. The array hashing algorithm, skips an increasing proportion of elements when the number of levels grows, so it's better for large pools, but when adding a new level it would be tricky to compute the new hash without rehashing everything from scratch. Maybe it's better in practice -- benchmarking is probably needed.

Reply:
If we define `isequal` we must define `hash`. Also your approach is OK I think. I am looking at the PR linearly, but I would lazily store the hash value so that it is not recomputed each time if it was already computed.

Reply:
Yeah I actually came to the same conclusion. Though that still leaves the option between hashing all levels and hashing only some of them (like arrays).

Reply:
I would hash all levels as you proposed (if we do it lazily this should not be that prohibitive).

Reply:
The point is that for arrays the assumption is that it is unlikely that they might differ in only one entry and have all other entries equal. However, this is quite possible for levels.